### PR TITLE
Enable usage of artifact-repository in composer-dev template

### PR DESCRIPTION
**Pull request by mokshasoul**
This enables `composer-dev` environments to use private pip repositories from GCP `artifact-registry`.

### CODE REVIEW

File: composer_local_dev/docker_files/entrypoint.sh [hunk 1]
```diff
@@ -28,6 +29,15 @@ if [ -f /var/local/setup_python_command.sh ]; then
     /var/local/setup_python_command.sh
 fi
 
+
+pip3 install keyring keyrings.google-artifactregistry-auth
```
Comment (ebartels) [inline composer_local_dev/docker_files/entrypoint.sh:33]: Should we only install keyring if a private index is used?

File: README.md [hunk 2]
```diff
@@ -454,3 +454,16 @@ You can use one of the following solutions:
     `~/Library/Group\ Containers/group.com.docker/settings.json` file and
     add `/opt` to `filesharingDirectories`.
 
+
```
Comment (piotrwrotniak) [inline README.md:457]: Let's add a section named 

`Configure private PyPI repository`
```
To allow installation of PyPI packages from private repositories define PIP_EXTRA_INDEX_URL environment variable that should contain an [alternate package index location](https://mothergeo-py.readthedocs.io/en/latest/development/how-to/alternate-pypi.html#how-to-use-an-alternate-pypi-package-index), eg.

PIP_EXTRA_INDEX_URL=http://pypi.mydomain.com:8080
```
after section `### Install or remove PyPI packages`

File: composer_local_dev/docker_files/entrypoint.sh (hunk 1 above)
Comment (piotrwrotniak) [reply composer_local_dev/docker_files/entrypoint.sh:33]: +1, please move this so it's executed only if needed.

File: composer_local_dev/docker_files/entrypoint.sh [hunk 3]
```diff
@@ -28,6 +29,15 @@ if [ -f /var/local/setup_python_command.sh ]; then
...
+if [ -z "${PRIVATE_PIP_INDEX}" ] 
+then
+  echo "no private index set"
+else
+  echo "index is set to ${PRIVATE_PIP_INDEX}"
```
Comment (piotrwrotniak) [inline composer_local_dev/docker_files/entrypoint.sh:38]: Please change to: `Adding private PyPI repository index: ${PIP_EXTRA_INDEX_URL}`

**Review by piotrwrotniak — CHANGES_REQUESTED** [review]
Thanks for the PR, please see my comments and please make the PR point to the development branch (we will resync this to main afterwards).

**Comment by mokshasoul** [comment]
> Thanks for the PR, please see my comments and please make the PR point to the development branch (we will resync this to main afterwards).

Hey @piotrwrotniak , just switched the branch as well, didn't see this comment thx for reviewing this

File: composer_local_dev/docker_files/entrypoint.sh (hunk 1 above)
Comment (mokshasoul) [reply composer_local_dev/docker_files/entrypoint.sh]: My comments seem to have disappeared but this is addressed within [changes to this commit](https://github.com/GoogleCloudPlatform/composer-local-dev/pull/21/commits/45a90d27ffc3b7d554c11924d0d91bb00eeb3872)

File: composer_local_dev/docker_files/entrypoint.sh [hunk 4]
```diff
@@ -28,6 +29,15 @@ if [ -f /var/local/setup_python_command.sh ]; then
     /var/local/setup_python_command.sh
 fi
 
+
+pip3 install keyring keyrings.google-artifactregistry-auth
+if [ -z "${PRIVATE_PIP_INDEX}" ] 
+then
+  echo "no private index set"
```
[the remark this answers is not in the document]
Comment (mokshasoul) [reply composer_local_dev/docker_files/entrypoint.sh]: My comments seem to have disappeared but this is addressed within [changes to this commit](https://github.com/GoogleCloudPlatform/composer-local-dev/pull/21/commits/45a90d27ffc3b7d554c11924d0d91bb00eeb3872)

File: composer_local_dev/docker_files/entrypoint.sh (hunk 3 above)
Comment (mokshasoul) [reply composer_local_dev/docker_files/entrypoint.sh]: My comments seem to have disappeared but this is addressed within [changes to this commit](https://github.com/GoogleCloudPlatform/composer-local-dev/pull/21/commits/45a90d27ffc3b7d554c11924d0d91bb00eeb3872)

File: README.md (hunk 2 above)
Comment (mokshasoul) [reply README.md]: [Addressed in this commit](https://github.com/GoogleCloudPlatform/composer-local-dev/pull/21/commits/6e3c453508299a1d3980cceea0cd170db33891fa)

**Comment by aflansburg** [comment]
Noticed this is stale-ish :( 
The change proposed by this PR would be most welcome. We had a [similar solution](https://github.com/brightsidehealth/composer-local-dev/blob/aed169dc40066ae142d6c11b52b02f15341913a2/composer_local_dev/docker_files/entrypoint.sh#L32-L49C3):

```sh
# added to composer_local_dev/docker_files/entrypoint.sh
sudo chown airflow:airflow .config

if [ -n "${PRIVATE_INDEX_URLS}" ]; then
  echo "Adding private PyPI repository indexes: ${PRIVATE_INDEX_URLS}"
  pip3 install keyring keyrings.google-artifactregistry-auth

  # Split PRIVATE_INDEX_URLS by comma and create a space-separated string of URLs
  OLDIFS=$IFS
  IFS=','
  # shellcheck disable=SC2086
  set -- $PRIVATE_INDEX_URLS
  IFS=$OLDIFS
  EXTRA_INDEX_URLS=""
  for url; do
    EXTRA_INDEX_URLS="${EXTRA_INDEX_URLS} ${url}"
  done
  # Export the URLs as an environment variable
  export PIP_EXTRA_INDEX_URL="${EXTRA_INDEX_URLS}"
  printf 'PIP_EXTRA_INDEX_URL="%s"\n' "${PIP_EXTRA_INDEX_URL}"
fi
```

And adding our `PRIVATE_INDEX_URLS` in the generated `composer/local-airflow-dev/variables.env` unquoted and comma delimited seemed to work well when using multiple private packages in Artifact registry.

File: composer_local_dev/docker_files/entrypoint.sh (hunk 3 above)
Comment (blocknomad) [reply composer_local_dev/docker_files/entrypoint.sh]: Extra index URLs are unsafe. Would you guys be interested in adding support for PIP_INDEX_URL as well? Thank you

**Comment by mokshasoul** [comment]
x

**Comment by tebesfinwo** [comment]
I have had success by leveraging environment variables via the `variables.env`. Environment variables I used:

```
PIP_EXTRA_INDEX_URL=YOUR_ARTIFACT_REPOSITORY
PIP_TRUSTED_HOST=us-python.pkg.dev
PIP_PREFER_BINARY=1
```

**Pull request closed without merging**